p2p/transport/webrtc/connection.go

```diff
@@ -0,0 +1,250 @@
+
+		accept: accept,
+	}
+
+	pc.OnDataChannel(func(dc *webrtc.DataChannel) {
+		log.Debugf("[%s] incoming datachannel: %s", localPeer, dc.Label())
+		id := *dc.ID()
+		var stream *dataChannel
+		dc.OnOpen(func() {
+			// datachannel cannot be detached before opening
+			rwc, err := dc.Detach()
+			if err != nil {
+				log.Errorf("[%s] could not detch channel: %s", localPeer, dc.Label())
+				return
+			}
+			stream = newDataChannel(dc, rwc, pc, nil, nil)
+			conn.addStream(id, stream)
+			accept <- stream
+		})
+
+		dc.OnClose(func() {
+			stream.remoteClosed()
+			conn.removeStream(id)
+		})
+	})
+
+	pc.OnConnectionStateChange(func(state webrtc.PeerConnectionState) {
+		if state == webrtc.PeerConnectionStateClosed || state == webrtc.PeerConnectionStateDisconnected {
+			conn.Close()
+		}
+	})
+
+	return conn
+}
+
+// ConnState implements transport.CapableConn
+func (c *connection) ConnState() network.ConnectionState {
+	return network.ConnectionState{}
+}
+
+// Implement network.MuxedConn
+
+func (c *connection) Close() error {
+	if c.IsClosed() {
+		return nil
+	}
+
+	c.scope.Done()
+	// cleanup routine
+	for _, stream := range c.streams {
+		_ = stream.Close()
+	}
+	c.cancel()
+	_ = c.pc.Close()
+	return nil
+}
+
+func (c *connection) IsClosed() bool {
+	select {
+	case <-c.ctx.Done():
+		return true
+	default:
+	}
+	return false
+}
+
+func (c *connection) OpenStream(ctx context.Context) (network.MuxedStream, error) {
+	if c.IsClosed() {
+		return nil, os.ErrClosed
+	}
+
+	result := make(chan struct {
+		network.MuxedStream
+		error
+	})
+	dc, err := c.pc.CreateDataChannel("", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	streamID := *dc.ID()
+	var stream *dataChannel
+	dc.OnOpen(func() {
+		rwc, err := dc.Detach()
+		if err != nil {
+			result <- struct {
+				network.MuxedStream
+				error
+			}{nil,
+				errDatachannel("could not detach", err),
+			}
+			return
+		}
+		stream = newDataChannel(dc, rwc, c.pc, nil, nil)
+		c.addStream(streamID, stream)
+		result <- struct {
+			network.MuxedStream
+			error
+		}{stream, err}
+	})
+
+	dc.OnClose(func() {
+		stream.remoteClosed()
+		c.removeStream(streamID)
+	})
+
+	select {
+	case <-ctx.Done():
+		_ = dc.Close()
+		return nil, ctx.Err()
+	case r := <-result:
+		return r.MuxedStream, r.error
+	}
+}
+
+func (c *connection) AcceptStream() (network.MuxedStream, error) {
+	select {
+	case <-c.ctx.Done():
+		return nil, os.ErrClosed
+	case stream := <-c.accept:
+		return stream, nil
+	}
+}
+
+// implement network.ConnSecurity
+func (c *connection) LocalPeer() peer.ID {
+	return c.localPeer
+}
+
+// only used during setup
+func (c *connection) setRemotePeer(id peer.ID) {
+	c.remotePeer = id
+}
+
+func (c *connection) LocalPrivateKey() ic.PrivKey {
+	return c.privKey
+}
+
+func (c *connection) RemotePeer() peer.ID {
+	return c.remotePeer
+}
+
+func (c *connection) RemotePublicKey() ic.PubKey {
+	return c.remoteKey
+}
+
+func (c *connection) setRemotePublicKey(key ic.PubKey) {
```

Comment:
Please refer to https://github.com/libp2p/go-libp2p/pull/1655/files#diff-097ca3a9a3a76a843b598d8b9b5f4ca582468b84894c7dfd2a976fd7135e88d2R277 . We create the connection object before we know the remote's Peer ID and public key. This is required since there are cases where a new datachannel may be initiated by the remote before callbacks to handle the datachannel are set up.

Reply:
I've removed this setter, but as mentioned above, we need a setter for the peer ID.